docs/design/2018-07-19-row-format.md

```diff
@@ -0,0 +1,88 @@
+
+- Author(s):     Yuxing ZHOU
+- Last updated:  2018-07-19
+
+## Abstract
+
+This proposal proposes a new row format for efficient decoding.
+
+## Background
+
+In the current row format, each column is encoded with the column ID followed by the column value and multiple columns are concatenated together. In order to find a specific column in the row, we need to start from the beginning and decode each column ID and the length of each value until we find the matched column. This decoding method is inefficient especially when the number of columns in a row is not small and the complexity is O(N).
+
+## Proposal
+
+The new format is defined as follows:
+
+![row-format](./row-format.jpeg)
+
+* 1 byte codec version
+
+    Its value starts from 128, so we can upgrade TiDB in a compatible way.
+
+* 1 byte row type flag containing the following attribute:
+
+	- 1 bit IsBig
+
+        A row is big if the max column ID is greater than 255 or the total size of the values is greater than 65535. We can add more types in this byte later if needed, for example storing 2 bytes column ID.
+
+* 2-byte number of not-null columns
+
+* 2-byte number of null columns
+
+    We can store 65536 columns in a row and this number is much larger than the upper limit in TiDB.
+
+* not-null column ID array
+
+    If there are too many columns, each column ID takes 4 bytes, otherwise the column ID takes 1 byte. The IDs array is sorted in an ascending order.
+
+* null column ID array
+
+    This array lists the IDs of all the columns with value `null`. The IDs array is sorted in an ascending order.
+
+* offset array
+
+     Specifies the end offset of each not-null column value, relative to the start of the first column value. If the row is big, one offset takes 4 bytes, otherwise it takes 2 bytes.
+
+* column values
+
+    For integer values, we store {byte, uint16, uint32, uint64} instead of varint for fast decoding. The type can be inferred by the column value size.
+The string value is stored as it is. Other types are encoded in the old way without the first type flag byte.
+
+
+To find a column in a row, we can perform a binary search on the not-null column ID array. If it’s not found, then we perform a binary search on the null column ID array.
+The complexity is O(logN).
+
+If the column ID can not be found in the row, we use the column default value in the schema.
+
+## Rationale
+
+The new format has better read performance but we need to consider the space efficiency.
+
+In most of the cases, column IDs are less than 255 and the total length of the values is less than 64KB, so we only compare the small row type with the old format.
+
+In the current row format, a string type column takes 1 (varint flag) + 1(colID) + 1 (bytes flag) + 1(varint len) + len(val). In the new row format, a string type column takes 1(colID) + 2(offset) + len(val), one byte less. So the string type is more space efficient and varint is more space efficient.
```

Comment:
Is "In the new row format, a string type column takes 1(colID) + 2(offset) + len(val), one byte less compared with the old row format." better and clear?

Reply:
yes